AC-742 refactoring code for better support for localization

LuGO0 opened the pull request:
## Description of what I changed
1. refactoring and extracting string resources for better support of localization in the client.
2. updated the ApplicationConstants


## Issue I worked on
JIRA Issue: https://issues.openmrs.org/projects/AC/issues/AC-742

## Checklist: I completed these to help reviewers :)
- [ ] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [x] I have **added tests** to cover my changes. (If you refactored
existing code that was well tested you do not have to add tests)

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

LuGO0 commented:
@f4ww4z  Sir please review the strings I have added to the file resources file after which I will add the translations for them!!.

LuGO0 commented on openmrs-client/schemas/org.openmrs.mobile.databases.AppDatabase/1.json:
```diff
@@ -554,7 +536,7 @@
+      "INSERT OR REPLACE INTO room_master_table (id,identity_hash) VALUES(42, \"2ca31877683087a2cef9a670e3bed3be\")"
     ]
   }
 }
```
I will remove this file from the commit when the changes are approved.

jawidMuhammadi commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/dialog/CameraOrGalleryPickerDialog.java:
```diff
@@ -49,7 +49,7 @@ public static CameraOrGalleryPickerDialog getInstance(DialogInterface.OnClickLis
 
 
-        String[] textResources = {"Take a photo", "Choose another", "Remove photo"};
+        String[] textResources = {getString(R.string.take_a_photo_alert_dialog_option), getString(R.string.choose_another_photo_alert_dialog_message), getString(R.string.remove_photo_alert_dialog_message)};
```
Instead of this long line of code, you could simply define these three elements as a `string array` inside `strings.xml` file then just use it here.

LuGO0 replied:
Yeah thats a good option thanks!!

jawidMuhammadi commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/formdisplay/FormDisplayMainPresenter.java:
```diff
@@ -118,8 +121,7 @@ public void createEncounter() {
                 new PatientDAO().updatePatient(mPatient.getId(),mPatient);
-                ToastUtil.error("Patient not yet registered. Form data is saved locally " +
-                        "and will sync when internet connection is restored. ");
+                ToastUtil.error(Resources.getSystem().getString(R.string.patient_not_resistered_error_message)+Resources.getSystem().getString(R.string.form_data_will_be_synced_later_no_internet_connection_error_message));
```
Why we are dividing this error message into different strings?

LuGO0 replied:
Because I could use the other substring somewhere else I think Ill have to find that out !!

jawidMuhammadi replied:
Ook, then try to rename the strings to a concise and short name, especially this string: `form_data_will_be_synced_later_no_internet_connection_error_message`

f4ww4z replied:
@LuGO0 I think this should be just 1 string resource, and shorten the string name.

LuGO0 replied:
@f4ww4z I will refactor it that way !!

jawidMuhammadi commented on openmrs-client/src/main/res/values/strings.xml:
```diff
@@ -376,4 +376,105 @@
+	<string name="boolean_flag">flag</string>
+	<string name="error_report">error</string>
+	<string name="package_manager">Package Manager</string>
+	<string name="google_API_key_token">com.google.android.geo.API_KEY</string>
```
As you are using this string as a constant string then definitely add the `translatable = false` attribute like this: `  <string name="google_API_key_token" translatable="false">com.google.android.geo.API_KEY</string>`

jawidMuhammadi replied:
Add this attribute to all your added strings which are not related to UI and do not need any translation.

LuGO0 replied:
Yeah sure I would do that :)

f4ww4z replied:
> Add this attribute to all your added strings which are not related to UI and do not need any translation.

We write these in `ApplicationConstants` anyway.

f4ww4z commented on openmrs-client/src/main/res/layout/fragment_login.xml:
```diff
@@ -74,7 +74,7 @@
 						android:layout_height="wrap_content"
 						android:layout_marginBottom="20dp"
-						android:hint="Url Address"
+						android:hint="@string/url_address"
```
See the string `login_url_hint`

LuGO0 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/addeditpatient/AddEditPatientFragment.java:
```diff
@@ -316,7 +315,7 @@ private Patient updatePatientWithData(Patient patient) {
 
         // Add gender
-        String[] genderChoices = {"M", "F"};
+        String[] genderChoices = {getString(R.string.male_gender_indicator), getString(R.string.female_gender_indicator)};
```
okay I will do that :)

LuGO0 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/dashboard/DashboardFragment.java:
```diff
@@ -82,8 +82,8 @@ private void showOverlayTutorialOne() {
         Target viewTarget = new ViewTarget(R.id.findPatientView, this.getActivity());
         new ShowcaseView.Builder(this.getActivity())
                 .setTarget(viewTarget)
-                .setContentTitle("Find Patients")
```
yeah I have it now resolved !!

LuGO0 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/dialog/CameraOrGalleryPickerDialog.java:
```diff
@@ -49,7 +49,7 @@ public static CameraOrGalleryPickerDialog getInstance(DialogInterface.OnClickLis
 
-        String[] textResources = {"Take a photo", "Choose another", "Remove photo"};
+        String[] textResources = {getString(R.string.take_a_photo_alert_dialog_option), getString(R.string.choose_another_photo_alert_dialog_message), getString(R.string.remove_photo_alert_dialog_message)};
         int[] imageResources = {R.drawable.ic_photo_camera, R.drawable.ic_photo_library, R.drawable.ic_photo_delete};
```
resolved

LuGO0 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/lastviewedpatients/LastViewedPatientsFragment.java:
```diff
@@ -71,7 +71,7 @@ public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                 mAdapter.finishActionMode();
             }else {
-                ToastUtil.error("No Internet Connection");
+                ToastUtil.error(getString(R.string.no_internet_connection_message));
```
it has a different message @f4ww4z though its related should I use that here??

LuGO0 commented on openmrs-client/src/main/java/org/openmrs/mobile/services/ConceptDownloadService.java:
```diff
@@ -89,8 +89,8 @@ private void showNotification(int downloadedConcepts) {
-                .setContentTitle("Downloading Concepts")
-                .setTicker("OpenMRS Android Client")
+                .setContentTitle(getString(R.string.downloading_concepts_notification_message))
+                .setTicker(getString(R.string.accessibility_services_app_name_message))
```
its OpenMRS Client on line16 strings.xml @f4ww4z sir will it do ??

f4ww4z replied:
Yes but you should change the string to `OpenMRS Android Client` . 

> its OpenMRS Client on line16 strings.xml @f4ww4z sir will it do ??

LuGO0 commented:
@f4ww4z  I refactored and I removed unused resources so all these resources got deleted when I thought only the unused strings would be deleted.
Is it fine or should I revert and force push changes again ??

LuGO0 commented:
@f4ww4z can you review again so that I can finally add translations to the added string resources!!

LuGO0 commented:
@f4ww4z any updates on this ??

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/databases/entities/VisitEntity.java:
```diff
@@ -3,6 +3,8 @@
 import androidx.room.ColumnInfo;
 import androidx.room.Entity;
+
+import org.openmrs.mobile.databases.tables.VisitTable;
```
No need for this change. Please pull the latest changes from the master branch.

f4ww4z commented on openmrs-client/src/main/java/org/openmrs/mobile/databases/entities/ConceptEntity.java:
```diff
@@ -15,6 +15,7 @@
 
 import androidx.room.PrimaryKey;
 import org.jetbrains.annotations.NotNull;
+import org.openmrs.mobile.databases.tables.ConceptTable;
```
No need for this change. Please pull the latest changes from the master branch.

rishabh-997 commented:
> some translations were tough I had to use google Translate XD !!
> @rishabh-997 please review the translations.
> @f4ww4z please review the changes I will squash my commits .

The Hindi translations are appropriate. Make sure to add JIRA ticket to the title and resolve the ci errors.

LuGO0 commented:
merge conflicts resolved and the build works fine for me idk why the CI build fails

rishabh-997 commented:
> merge conflicts resolved and the build works fine for me idk why the CI build fails

You can click on CI details to see which particular test is failing and then run that test locally to check for exact error... You can run the entire presenter folder all together to see if all the tests are passing or not.

f4ww4z commented:
@LuGO0 Please fix the merge conflicts and also take a look at these [null pointer exceptions in the CI](https://travis-ci.org/github/openmrs/openmrs-contrib-android-client/builds/685180368#L944).

LuGO0 commented:
okay @rishabh-997  Im looking into CI errors  btw the build works fine for me!!

LuGO0 commented:
> @LuGO0 Please fix the merge conflicts and also take a look at these [null pointer exceptions in the CI](https://travis-ci.org/github/openmrs/openmrs-contrib-android-client/builds/685180368#L944).

@f4ww4z  I am looking into these errors !! and will resolve conflicts

LuGO0 commented:
@f4ww4z I found out that`Resources.getSystem()` is null in the tests bt it works fine when app executes should I make changes in the testFile ?

f4ww4z commented:
> @f4ww4z I found out that`Resources.getSystem()` is null in the tests bt it works fine when app executes should I make changes in the testFile ? I will resolve conflicts once these tests pass !

@LuGO0 Yes you need to change test files. You can either use mocks to mock the `getString` part or create instrumentation tests.

LuGO0 commented:
Okay @f4ww4z

LuGO0 commented:
![Capture](https://user-images.githubusercontent.com/30355336/82160060-35a6a080-98b0-11ea-8b30-0ebe54d97235.PNG)

@rishabh-997 @f4ww4z  can you help me with this last test I couldnt fix it with mocking !!

f4ww4z commented:
@LuGO0 can you commit your current test changes so we can take a look?

LuGO0 commented:
Yes @f4ww4z I did force push its in the added translations commit

LuGO0 commented on openmrs-client/src/test/java/org/openmrs/mobile/test/presenters/PatientDashboardVisitsPresenterTest.java:
```diff
@@ -83,38 +86,41 @@ public void shouldShowPatientsVisitsOnStartUp_allOK(){
         verify(view).showStartVisitDialog(false);
     }
 
     @Test
```
specifically this test fails and due on line-111, due to differences between expected and actual params I couldnt figure out what!!

rishabh-997 replied:
This test seems to work fine in the CI results below...

rishabh-997 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/formlist/FormListPresenter.java:
```diff
@@ -89,17 +88,17 @@ public void listItemClicked(int position, String formName) {
             view.startFormDisplayActivity(formName, patientId, valueRefString, encounterType);
         } else {
-            view.showError("There is no encounter type called " + formName);
+            view.showError(Resources.getSystem().getString(R.string.no_such_form_name_error_message, formName));
```
`Resources.getSystem()` references to the system resources and might cause a crash if used incorrectly. Here take a look at this [post](https://stackoverflow.com/questions/8535626/notfoundexception-when-looking-for-drawable-resource)... Using this is not generally a good practice as it causes `NullPointerException` and this is what exactly happened in you test which received `NULL` value instead of string...

LuGO0 replied:
@rishabh-997  yeah inside the UnitTests this was a problem. 8 tests were failing before due to this reason of Resources not being a part of JVM as the tests are run in JVM and Resources is an Android specifc class but after I mocked this using powerMock 7 tests passed now only this one remains and I dont get why mocking dosent help here ?

LuGO0 replied:
the test case passes @rishabh-997  it was really a silly mistake I was not giving the correct params to getString() it expected 2 I was provinding only 1!!

LuGO0 commented:
@f4ww4z  please review !!

rishabh-997 commented:
The app crashes at Login Page with the following error
```
Process: org.openmrs.mobile, PID: 14568
    android.content.res.Resources$NotFoundException: String resource ID #0x7f1100c4
        at android.content.res.Resources.getText(Resources.java:375)
        at android.content.res.Resources.getString(Resources.java:468)
        at org.openmrs.mobile.activities.login.LoginPresenter$2.onResponse(LoginPresenter.java:240)
```
The reason seems to be the `nullPointerException` due to `Resources.getSystem()`

LuGO0 commented:
> The app crashes at Login Page with the following error
> 
> ```
> Process: org.openmrs.mobile, PID: 14568
>     android.content.res.Resources$NotFoundException: String resource ID #0x7f1100c4
>         at android.content.res.Resources.getText(Resources.java:375)
>         at android.content.res.Resources.getString(Resources.java:468)
>         at org.openmrs.mobile.activities.login.LoginPresenter$2.onResponse(LoginPresenter.java:240)
> ```
> 
> The reason seems to be the `nullPointerException` due to `Resources.getSystem()`

@rishabh-997  I am looking into this now

LuGO0 commented:
@rishabh-997  @f4ww4z  I have removed all instances of `Resources.getSystem()` from all the presenters and instead delegated the views those are attached to it to display the strings.
but what should I do with the Repository class which uses `ToastUtil.error ` to toast a lot of errors I couldnt find a work around for repository classes

LuGO0 commented:
> @rishabh-997 @f4ww4z I have removed all instances of `Resources.getSystem()` from all the presenters and instead delegated the views those are attached to it to display the strings.
> but what should I do with the Repository class which uses `ToastUtil.error ` to toast a lot of errors I couldnt find a work around for repository classes

the best I could think was a callback interface which woudl be implemented in the view and delegated down uptill the repository class!!

LuGO0 commented:
@f4ww4z  apart from the patientRepository I refactored all hard coded string and for repository section as well I have added resources and made a comment over its usage . please review !!

LuGO0 commented:
@rishabh-997 I coudnt test it on my phone how do we log in when demo server is down is there any way ??

rishabh-997 commented:
> @rishabh-997 I coudnt test it on my phone how do we log in when demo server is down is there any way ??

You can use alternative [server](https://qa-refapp.openmrs.org/openmrs/).

LuGO0 commented:
> > @rishabh-997 I coudnt test it on my phone how do we log in when demo server is down is there any way ??
> 
> You can use alternative [server](https://qa-refapp.openmrs.org/openmrs/).

thanks Rishabh !!

LuGO0 commented:
@f4ww4z I have run all the tests locally now it works fine and the app also runs fine as far as I have tested please review !!

rishabh-997 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/dialog/CameraOrGalleryPickerDialog.java:
```diff
@@ -62,6 +64,15 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
+    //     AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
+    //     builder.setAdapter(new GalleryOrCameraPickerListAdapter(getActivity(),
+    //         R.layout.list_gallery_or_camera_item, R.id.textView, R.id.imageView,
+    //         textResources, imageResources), listener);
```
Why are all these comments???

rishabh-997 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/formdisplay/FormDisplayMainPresenter.java:
```diff
@@ -115,19 +111,18 @@ public void createEncounter() {
-                ToastUtil.error("Patient not yet registered. Form data is saved locally " +
-                        "and will sync when internet connection is restored. ");
+                new PatientDAO().updatePatient(mPatient.getId(), mPatient);
+                mFormDisplayView.showToast();
```
Or can we pass an ApplicationContext to the presenter to get System Resources instead of creating another function??? What's your say on this @f4ww4z sir???

LuGO0 replied:
@rishabh actually its a bad practice to get context to presenters so I tried to avoid it as far as i can

rishabh-997 commented on openmrs-client/src/main/java/org/openmrs/mobile/activities/login/LoginFragment.java:
```diff
@@ -361,14 +359,14 @@ public void startFormListService() {
-                        mUrl.selectAll();
-                    })
-                    .show();
+            createSnackbar(getString(R.string.failed_fetching_servers_location))
```
Don't make the same message displayed as it is categorized into:
1. Invalid URL
2. Server Down

rishabh-997 commented on openmrs-client/src/main/java/org/openmrs/mobile/api/RestServiceBuilder.java:
```diff
@@ -30,36 +30,31 @@
-                        .method(original.method(), original.body());
-
+                    .header("Authorization", basic)
+                    .header("Accept", "application/json")
```
Should this be added in Application Constant???

LuGO0 replied:
should I add them @rishabh-997 ???

LuGO0 replied:
I can add them if you think that way @rishabh-997 !!

rishabh-997 replied:
I don't think this won't be necessary as target is to support better localization...

LuGO0 replied:
Ohkk @rishabh-997 resolving it

LuGO0 commented:
@rishabh I was doing a code parsing for all the hard coded string and it did code reformatting I dont know how bt It was good so I let it that way.
I will take care of this from next time

LuGO0 commented:
@f4ww4z @rishabh-997 please further review it .Its getting difficult to merge conflicts with this PR and keep it upto date with master 😅

f4ww4z commented on openmrs-client/src/test/java/org/openmrs/mobile/utilities/DateUtilsTest.java:
```diff
@@ -36,150 +38,146 @@
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotSame;
-import static org.junit.Assert.assertNull;
+import static org.junit.Assert.*;
```
Don't change to star imports.

LuGO0 replied:
fixed this @f4ww4z Sir.

f4ww4z commented on openmrs-client/src/test/java/org/openmrs/mobile/utilities/DateUtilsTest.java:
```diff
@@ -36,150 +38,146 @@
-        DateTime dateTime;
-        String nullDateAsString = null;
-        DateTimeFormatter formatter = DateTimeFormat
+    @RunWith(PowerMockRunner.class)
```
No need to reindent this.

LuGO0 replied:
done @f4ww4z Sir.

LuGO0 commented:
@f4ww4z Sir I have made requested changes and will sqash changes once you approve of them!!
